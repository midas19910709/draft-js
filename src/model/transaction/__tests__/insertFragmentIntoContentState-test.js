/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+ui_infra
 * @format
 */

'use strict';

jest.disableAutomock();

jest.mock('generateRandomKey');

const BlockMapBuilder = require('BlockMapBuilder');
const CharacterMetadata = require('CharacterMetadata');
const ContentBlock = require('ContentBlock');
const Immutable = require('immutable');

const getSampleStateForTesting = require('getSampleStateForTesting');
const insertFragmentIntoContentState = require('insertFragmentIntoContentState');

const {contentState, selectionState} = getSampleStateForTesting();
const {EMPTY} = CharacterMetadata;
const {List, Map} = Immutable;

const DEFAULT_BLOCK_CONFIG = {
  key: 'j',
  type: 'unstyled',
  text: 'xx',
  characterList: List.of(EMPTY, EMPTY),
  data: new Map({a: 1}),
};

const initialBlock = contentState.getBlockMap().first();

const createFragment = (fragment = {}) => {
  fragment = Array.isArray(fragment) ? fragment : [fragment];

  return BlockMapBuilder.createFromArray(
    fragment.map(
      config =>
        new ContentBlock({
          ...DEFAULT_BLOCK_CONFIG,
          ...config,
        }),
    ),
  );
};

const assertInsertFragmentIntoContentState = (
  fragment,
  selection = selectionState,
  content = contentState,
) => {
  expect(
    insertFragmentIntoContentState(content, selection, fragment).toJS(),
  ).toMatchSnapshot();
};

test('must throw if no fragment is provided', () => {
  const fragment = BlockMapBuilder.createFromArray([]);
  expect(() => {
    insertFragmentIntoContentState(contentState, selectionState, fragment);
  }).toThrow();
});

test('must apply fragment to the start', () => {
  assertInsertFragmentIntoContentState(createFragment());
});

test('must apply fragment to within block', () => {
  assertInsertFragmentIntoContentState(
    createFragment(),
    selectionState.merge({
      focusOffset: 2,
      anchorOffset: 2,
      isBackward: false,
    }),
  );
});

test('must apply fragment at the end', () => {
  assertInsertFragmentIntoContentState(
    createFragment(),
    selectionState.merge({
      focusOffset: initialBlock.getLength(),
      anchorOffset: initialBlock.getLength(),
      isBackward: false,
    }),
  );
});

test('must apply multiblock fragments', () => {
  assertInsertFragmentIntoContentState(
    createFragment([
      DEFAULT_BLOCK_CONFIG,
      {
        key: 'k',
        text: 'yy',
        data: new Map({b: 2}),
      },
    ]),
  );
});
