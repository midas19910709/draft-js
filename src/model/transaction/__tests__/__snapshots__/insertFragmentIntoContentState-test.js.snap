// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must apply fragment at the end 1`] = `
Object {
  "blockMap": Object {
    "a": Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
      ],
      "data": Object {
        "a": 1,
      },
      "depth": 0,
      "key": "a",
      "text": "Alphaxx",
      "type": "unstyled",
    },
    "b": Object {
      "characterList": Array [
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
      ],
      "data": Object {},
      "depth": 0,
      "key": "b",
      "text": "Bravo",
      "type": "unordered-list-item",
    },
    "c": Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
      ],
      "data": Object {},
      "depth": 0,
      "key": "c",
      "text": "Test",
      "type": "code-block",
    },
    "d": Object {
      "characterList": Array [],
      "data": Object {},
      "depth": 0,
      "key": "d",
      "text": "",
      "type": "code-block",
    },
    "e": Object {
      "characterList": Array [],
      "data": Object {},
      "depth": 0,
      "key": "e",
      "text": "",
      "type": "code-block",
    },
    "f": Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
      ],
      "data": Object {},
      "depth": 0,
      "key": "f",
      "text": "Charlie",
      "type": "blockquote",
    },
  },
  "entityMap": Object {},
  "selectionAfter": Object {
    "anchorKey": "a",
    "anchorOffset": 7,
    "focusKey": "a",
    "focusOffset": 7,
    "hasFocus": true,
    "isBackward": false,
  },
  "selectionBefore": Object {
    "anchorKey": "a",
    "anchorOffset": 5,
    "focusKey": "a",
    "focusOffset": 5,
    "hasFocus": true,
    "isBackward": false,
  },
}
`;

exports[`must apply fragment to the start 1`] = `
Object {
  "blockMap": Object {
    "a": Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
      ],
      "data": Object {
        "a": 1,
      },
      "depth": 0,
      "key": "a",
      "text": "xxAlpha",
      "type": "unstyled",
    },
    "b": Object {
      "characterList": Array [
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
      ],
      "data": Object {},
      "depth": 0,
      "key": "b",
      "text": "Bravo",
      "type": "unordered-list-item",
    },
    "c": Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
      ],
      "data": Object {},
      "depth": 0,
      "key": "c",
      "text": "Test",
      "type": "code-block",
    },
    "d": Object {
      "characterList": Array [],
      "data": Object {},
      "depth": 0,
      "key": "d",
      "text": "",
      "type": "code-block",
    },
    "e": Object {
      "characterList": Array [],
      "data": Object {},
      "depth": 0,
      "key": "e",
      "text": "",
      "type": "code-block",
    },
    "f": Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
      ],
      "data": Object {},
      "depth": 0,
      "key": "f",
      "text": "Charlie",
      "type": "blockquote",
    },
  },
  "entityMap": Object {},
  "selectionAfter": Object {
    "anchorKey": "a",
    "anchorOffset": 2,
    "focusKey": "a",
    "focusOffset": 2,
    "hasFocus": true,
    "isBackward": false,
  },
  "selectionBefore": Object {
    "anchorKey": "a",
    "anchorOffset": 0,
    "focusKey": "a",
    "focusOffset": 0,
    "hasFocus": true,
    "isBackward": false,
  },
}
`;

exports[`must apply fragment to within block 1`] = `
Object {
  "blockMap": Object {
    "a": Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
      ],
      "data": Object {
        "a": 1,
      },
      "depth": 0,
      "key": "a",
      "text": "Alxxpha",
      "type": "unstyled",
    },
    "b": Object {
      "characterList": Array [
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
      ],
      "data": Object {},
      "depth": 0,
      "key": "b",
      "text": "Bravo",
      "type": "unordered-list-item",
    },
    "c": Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
      ],
      "data": Object {},
      "depth": 0,
      "key": "c",
      "text": "Test",
      "type": "code-block",
    },
    "d": Object {
      "characterList": Array [],
      "data": Object {},
      "depth": 0,
      "key": "d",
      "text": "",
      "type": "code-block",
    },
    "e": Object {
      "characterList": Array [],
      "data": Object {},
      "depth": 0,
      "key": "e",
      "text": "",
      "type": "code-block",
    },
    "f": Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
      ],
      "data": Object {},
      "depth": 0,
      "key": "f",
      "text": "Charlie",
      "type": "blockquote",
    },
  },
  "entityMap": Object {},
  "selectionAfter": Object {
    "anchorKey": "a",
    "anchorOffset": 4,
    "focusKey": "a",
    "focusOffset": 4,
    "hasFocus": true,
    "isBackward": false,
  },
  "selectionBefore": Object {
    "anchorKey": "a",
    "anchorOffset": 2,
    "focusKey": "a",
    "focusOffset": 2,
    "hasFocus": true,
    "isBackward": false,
  },
}
`;

exports[`must apply multiblock fragments 1`] = `
Object {
  "blockMap": Object {
    "a": Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
      ],
      "data": Object {
        "a": 1,
      },
      "depth": 0,
      "key": "a",
      "text": "xx",
      "type": "unstyled",
    },
    "b": Object {
      "characterList": Array [
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
        Object {
          "entity": "1",
          "style": Array [
            "BOLD",
          ],
        },
      ],
      "data": Object {},
      "depth": 0,
      "key": "b",
      "text": "Bravo",
      "type": "unordered-list-item",
    },
    "c": Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
      ],
      "data": Object {},
      "depth": 0,
      "key": "c",
      "text": "Test",
      "type": "code-block",
    },
    "d": Object {
      "characterList": Array [],
      "data": Object {},
      "depth": 0,
      "key": "d",
      "text": "",
      "type": "code-block",
    },
    "e": Object {
      "characterList": Array [],
      "data": Object {},
      "depth": 0,
      "key": "e",
      "text": "",
      "type": "code-block",
    },
    "f": Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
        Object {
          "entity": null,
          "style": Array [
            "ITALIC",
          ],
        },
      ],
      "data": Object {},
      "depth": 0,
      "key": "f",
      "text": "Charlie",
      "type": "blockquote",
    },
    "key0": Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
        Object {
          "entity": null,
          "style": Array [],
        },
      ],
      "data": Object {
        "b": 2,
      },
      "depth": 0,
      "key": "key0",
      "text": "yyAlpha",
      "type": "unstyled",
    },
  },
  "entityMap": Object {},
  "selectionAfter": Object {
    "anchorKey": "key0",
    "anchorOffset": 2,
    "focusKey": "key0",
    "focusOffset": 2,
    "hasFocus": true,
    "isBackward": false,
  },
  "selectionBefore": Object {
    "anchorKey": "a",
    "anchorOffset": 0,
    "focusKey": "a",
    "focusOffset": 0,
    "hasFocus": true,
    "isBackward": false,
  },
}
`;
